hw3/02-charts.js

```diff
@@ -1,45 +1,47 @@
 const backgroundColors = [
-  'rgba(54, 162, 235, 0.8)',
-  'rgba(255, 206, 86, 0.8)',
-  'rgba(255, 99, 132, 0.8)',
-  'rgba(75, 192, 192, 0.8)',
-  'rgba(153, 102, 255, 0.8)',
-  'rgba(255, 159, 64, 0.8)',
-  'rgba(199, 199, 199, 0.8)',
-  'rgba(83, 102, 255, 0.8)',
-  'rgba(40, 159, 64, 0.8)',
-  'rgba(210, 199, 199, 0.8)',
-  'rgba(78, 52, 199, 0.8)',
+  "rgba(54, 162, 235, 0.8)",
+  "rgba(255, 206, 86, 0.8)",
+  "rgba(255, 99, 132, 0.8)",
+  "rgba(75, 192, 192, 0.8)",
+  "rgba(153, 102, 255, 0.8)",
+  "rgba(255, 159, 64, 0.8)",
+  "rgba(199, 199, 199, 0.8)",
+  "rgba(83, 102, 255, 0.8)",
+  "rgba(40, 159, 64, 0.8)",
+  "rgba(210, 199, 199, 0.8)",
+  "rgba(78, 52, 199, 0.8)",
 ];
 
 const borderColors = [
-  'rgba(54, 162, 235, 1)',
-  'rgba(255, 206, 86, 1)',
-  'rgba(255, 99, 132, 1)',
-  'rgba(75, 192, 192, 1)',
-  'rgba(153, 102, 255, 1)',
-  'rgba(255, 159, 64, 1)',
-  'rgba(159, 159, 159, 1)',
-  'rgba(83, 102, 255, 1)',
-  'rgba(40, 159, 64, 1)',
-  'rgba(210, 199, 199, 1)',
-  'rgba(78, 52, 199, 1)',
+  "rgba(54, 162, 235, 1)",
+  "rgba(255, 206, 86, 1)",
+  "rgba(255, 99, 132, 1)",
+  "rgba(75, 192, 192, 1)",
+  "rgba(153, 102, 255, 1)",
+  "rgba(255, 159, 64, 1)",
+  "rgba(159, 159, 159, 1)",
+  "rgba(83, 102, 255, 1)",
+  "rgba(40, 159, 64, 1)",
+  "rgba(210, 199, 199, 1)",
+  "rgba(78, 52, 199, 1)",
 ];
 
 // url for the Thrones API
-const url = 'https://thronesapi.com/api/v2/Characters';
+const url = "https://thronesapi.com/api/v2/Characters";
 
-const renderChart = () => {
-  const donutChart = document.querySelector('.donut-chart');
+const renderChart = async () => {
+  const donutChart = document.querySelector(".donut-chart");
+
+  const { familyNameList, familyCounters } = await extractAndAnalyzeData(url);
```

Comment:
Does this function "extractAndAnalyzeData()" actually return a promise? 

Why is "await" used here?

Reply:
yes it is, it has an `async` tag https://github.com/ptmdmusique/cs564-frontend-homework/pull/3/files/95501c5368eafc24100bedab119319318157ab38#diff-5e070198c5c4360d6f5c575fbfbf8e3ee8e31a1e310d360819445076b0e26379R75 

and also do 2 `await` inside https://github.com/ptmdmusique/cs564-frontend-homework/pull/3/files/95501c5368eafc24100bedab119319318157ab38#diff-5e070198c5c4360d6f5c575fbfbf8e3ee8e31a1e310d360819445076b0e26379R80-R81

Reply:
I had already looked at the entire file before making the comment. I just phrased it as a question so you could dig deeper into it.

Just because it has an "async" keyword does not mean that it automatically returns a promise. Just because the actual fetch API nested within it returns a promise and hence uses "await", does not mean the encompassing function returns a promise.

Reply:
> Just because it has an "async" keyword does not mean that it automatically returns a promise

I think it is. Even if it's not doing anything, the whole function still returns a promise and you can `await` it. Here is a [sandbox](https://www.typescriptlang.org/play?#code/MYewdgzgLgBFCm0CiBbADlAngQQpswAYgK4FQCW4MAvDAIZ4EwAUAlDQHwwDeMAvgChQkWAmTosJMpTA0W7al16Dh0OIii58RUsApVaDbfM48BMegHc65WGHiWYABQBOIFOQjxmzF4hAANgBu8ApK5hYwXlAAKuQo8CDEUD5hZpGRqoHwAHQBIADmzABEAJKQ5AAm8DBobh5exawA3BEZfhCBIcwAjABMAMwtbfwANDA9AAzTrBF8s4JC4GpisIaMwCaK6TAA9LswAOo1wHSy0OQBAVY2ogAWnhF01rbq4hg4G1J6MmytEfsYAA5EAUYA1KB3E7uNCXeAuGCWS7XawuWTECBwB6YigJJ4vUQaVAfb76MB-ASUiyAkFgiFQmD5AowEAuaoI+AhWSQpIFO5YmoAM10ZJglRAiDAAHJYH4oMQ0fRavVPPAIlkArkmSUAELwQWsmpGAhNVoWZ63N6aL4i37DCwarWFErYQUIBHG4CmgSLVZ-IA). Notice how TS reports the `Promise` as the return type

Reply:
If it's not a promise (no `async` flag), there is no way to `await` inside the function

Reply:
> > Just because it has an "async" keyword does not mean that it automatically returns a promise
> 
> I think it is. Even if it's not doing anything, the whole function still returns a promise and you can `await` it. Here is a [sandbox](https://www.typescriptlang.org/play?#code/MYewdgzgLgBFCm0CiBbADlAngQQpswAYgK4FQCW4MAvDAIZ4EwAUAlDQHwwDeMAvgChQkWAmTosJMpTA0W7al16Dh0OIii58RUsApVaDbfM48BMegHc65WGHiWYABQBOIFOQjxmzF4hAANgBu8ApK5hYwXlAAKuQo8CDEUD5hZpGRqoHwAHQBIADmzABEAJKQ5AAm8DBobh5exawA3BEZfhCBIcwAjABMAMwtbfwANDA9AAzTrBF8s4JC4GpisIaMwCaK6TAA9LswAOo1wHSy0OQBAVY2ogAWnhF01rbq4hg4G1J6MmytEfsYAA5EAUYA1KB3E7uNCXeAuGCWS7XawuWTECBwB6YigJJ4vUQaVAfb76MB-ASUiyAkFgiFQmD5AowEAuaoI+AhWSQpIFO5YmoAM10ZJglRAiDAAHJYH4oMQ0fRavVPPAIlkArkmSUAELwQWsmpGAhNVoWZ63N6aL4i37DCwarWFErYQUIBHG4CmgSLVZ-IA). Notice how TS reports the `Promise` as the return type

You were right about this. Async functions always return a promise. If the return value of an async function is not explicitly a promise, it will be implicitly wrapped in a promise.

However, even the MDN docs warn that: Even though the return value of an async function behaves as if it's wrapped in a Promise.resolve, they are not equivalent.

So I guess we both had our points.